File: php/src/Model/JoinLink.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Model;

/**
 * A QR join link. `maxUses` is null for unlimited; 0 would mean exhausted.
 *
 * Response model: built from a decoded JSON array and read-only thereafter.
 */
final class JoinLink
{
    public function __construct(
        public readonly string $id,
        public readonly string $label,
        /** 'active' | 'revoked' | 'expired' | 'exhausted' */
        public readonly string $status,
        public readonly ?string $teamId,
        /** @var list<string> */
        public readonly array $tags,
        public readonly bool $requiresApproval,
        /** Null means unlimited — 0 would mean exhausted. */
        public readonly ?int $maxUses,
        public readonly int $useCount,
        /** ISO-8601 */
        public readonly string $createdAt,
        /** ISO-8601 */
        public readonly string $expiresAt,
        public readonly ?string $revokedAt,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     * @internal
     */
    public static function fromArray(array $data): self
    {
        return new self(
            id: (string) ($data['id'] ?? ''),
            label: (string) ($data['label'] ?? ''),
            status: (string) ($data['status'] ?? ''),
            teamId: isset($data['teamId']) ? (string) $data['teamId'] : null,
            tags: \array_values(\array_map(strval(...), $data['tags'] ?? [])),
            requiresApproval: (bool) ($data['requiresApproval'] ?? false),
            maxUses: isset($data['maxUses']) ? (int) $data['maxUses'] : null,
            useCount: (int) ($data['useCount'] ?? 0),
            createdAt: (string) ($data['createdAt'] ?? ''),
            expiresAt: (string) ($data['expiresAt'] ?? ''),
            revokedAt: isset($data['revokedAt']) ? (string) $data['revokedAt'] : null,
        );
    }
}

File: java/src/main/java/io/fivexer/sdk/model/JoinLink.java
package io.fivexer.sdk.model;

import java.util.List;

/**
 * A QR join link. {@code maxUses} is null for unlimited; 0 would mean exhausted.
 *
 * <p>Populated by Gson via field reflection and exposed read-only; callers receive instances
 * from the client and never construct them.
 */
public class JoinLink {
    private String id;
    private String label;
    private String status;
    private String teamId;
    private List<String> tags;
    private Boolean requiresApproval;
    private Integer maxUses;
    private Integer useCount;
    private String createdAt;
    private String expiresAt;
    private String revokedAt;

    public String getId() { return id; }
    public String getLabel() { return label; }
    public String getStatus() { return status; }
    public String getTeamId() { return teamId; }
    public List<String> getTags() { return tags; }
    public Boolean getRequiresApproval() { return requiresApproval; }
    public Integer getMaxUses() { return maxUses; }
    public Integer getUseCount() { return useCount; }
    public String getCreatedAt() { return createdAt; }
    public String getExpiresAt() { return expiresAt; }
    public String getRevokedAt() { return revokedAt; }
}
